src/node_api_async.cc

```diff
@@ -1,69 +1,105 @@
 #include "node_api_async.h"
 #include "uv.h"
 
-typedef struct napi_work_impl__ {
-  uv_work_t* work;
-  void* data;
-  void (*execute)(void* data);
-  void (*complete)(void* data);
-  void (*destroy)(void* data);
-} napi_work_impl;
+namespace uvimpl {
 
-napi_work napi_create_async_work() {
-  napi_work_impl* worker =
-      reinterpret_cast<napi_work_impl*>(malloc(sizeof(napi_work_impl)));
-  uv_work_t* req = reinterpret_cast<uv_work_t*>(malloc(sizeof(uv_work_t)));
-  req->data = worker;
-  worker->work = req;
-  return reinterpret_cast<napi_work>(worker);
-}
+// Wrapper around uv_work_t which calls user-provided callbacks.
+class Work {
+ private:
+  explicit Work(napi_env env,
+                napi_async_callback execute = nullptr,
+                napi_async_callback complete = nullptr,
+                napi_async_callback destroy = nullptr,
+                void* data = nullptr)
+     : _env(env),
+     _data(data),
+     _execute(execute),
+     _complete(complete),
+     _destroy(destroy) {
+    _request = new uv_work_t;
+    _request->data = this;
+  }
 
-void napi_delete_async_work(napi_work w) {
-  napi_work_impl* worker = reinterpret_cast<napi_work_impl*>(w);
-  if (worker != NULL) {
-    if (worker->work != NULL) {
-      delete reinterpret_cast<uv_work_t*>(worker->work);
-    }
-    delete worker;
-    worker = NULL;
+  ~Work() { 
+	delete _request;
+  };
+
+ public:
+  static Work* New(napi_env env,
+                   napi_async_callback execute,
+                   napi_async_callback complete,
+                   napi_async_callback destroy,
+                   void* data) {
+    return new Work(env, execute, complete, destroy, data);
   }
-}
 
-void napi_async_set_data(napi_work w, void* data) {
-  napi_work_impl* worker = reinterpret_cast<napi_work_impl*>(w);
-  worker->data = data;
-}
+  static void Delete(Work* work) {
+    delete work;
+  }
 
-void napi_async_set_execute(napi_work w, void (*execute)(void* data)) {
-  napi_work_impl* worker = reinterpret_cast<napi_work_impl*>(w);
-  worker->execute = execute;
-}
+  static napi_status Queue(Work* work) {
+    // Consider: Choose a different loop based on the napi_env?
+    int result = uv_queue_work(uv_default_loop(),
```

Comment:
@gabrielschulhof Node actually does use non-default uv loops in some places. `uv_loop_new()` is an obsolete API; look for `uv_loop_init()` instead:

```
deps\uv\include\uv.h:260:UV_EXTERN int uv_loop_init(uv_loop_t* loop);
deps\uv\include\uv.h:265: *  allocate the loop manually and use uv_loop_init instead.
src\debug-agent.cc:77:  err = uv_loop_init(&child_loop_);
src\inspector_agent.cc:547:  int err = uv_loop_init(&child_loop_);
src\node_watchdog.cc:36:  rc = uv_loop_init(loop_);
src\spawn_sync.cc:466:  CHECK_EQ(uv_loop_init(uv_loop_), 0);
src\tracing\agent.cc:19:  int err = uv_loop_init(&tracing_loop_);
test\cctest\test_inspector_socket.cc:364:    GTEST_ASSERT_EQ(0, uv_loop_init(&loop));
test\cctest\test_inspector_socket_server.cc:74:    EXPECT_EQ(0, uv_loop_init(&loop));
```